torchao/prototype/low_bit_optim/cpu_offload.py

```diff
@@ -0,0 +1,101 @@
+from typing import Type
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class CPUOffloadOptimizer:
+    def __init__(self, params, optimizer_class: Type[Optimizer], *, offload_gradients: bool = False, **kwargs) -> None:
+        """Offload optimizer to CPU for single-GPU training. This will reduce GPU memory by the size of optimizer state.
+        Optimizer step will be done on CPU.
+
+        Args
+            params: a list of parameters or parameter groups.
+            optimizer_class: constructor of the base optimizer.
+            offload_gradients: free GPU gradients once they are moved to CPU. Not compatible with gradient accumulation.
+            kwargs: other keyword arguments to be passed to the base optimizer e.g. `lr`, `weight_decay`.
+        """
+        param_groups = list(params)
+        if len(param_groups) == 0:
+            raise ValueError("optimizer got an empty parameter list")
+        if not isinstance(param_groups[0], dict):
+            param_groups = [{"params": param_groups}]
+
+        self.param_cuda2cpu_map = dict()
+        self.optim_dict = dict()
+        self.stream = torch.cuda.Stream()
+
+        # the queue maintains the order which param we should do optim step on first.
+        self.queue = dict()
+
+        def backward_hook(p_cuda):
+            if p_cuda.grad is not None:
+                p_cpu = self.param_cuda2cpu_map[p_cuda]
+
+                # make sure backward for this param finishes
+                self.stream.wait_stream(torch.cuda.current_stream())
+                with torch.cuda.stream(self.stream):
+                    p_cpu.grad.copy_(p_cuda.grad, non_blocking=True)
+
+                # we rely on CPython implementation of dictionary, which preserves insertion order.
+                # if a param is added again (e.g. due to gradient accumulation), it is moved to the
+                # end of the queue by removing and inserting it again.
+                if p_cuda in self.queue:
+                    del self.queue[p_cuda]
+                self.queue[p_cuda] = self.stream.record_event()
+
+                # deallocate CUDA gradients once D2H transfer finishes.
+                if offload_gradients:
+                    p_cuda.grad.record_stream(self.stream)
```

Comment:
Note that `record_stream` will have non-deterministic memory behavior (namely, when the CUDA tensor gets freed depends on when its last GPU kernel finishes, which is difficult to reason about precisely).

We really want to move away from using `record_stream`, but it can make your implementation more complicated.

Reply:
Just curious, what are the alternatives?

Reply:
The crux of the issue is that `p_cuda.grad` is allocated in the default stream but has ops on it in a different stream, so there is a producer/consumer stream relationship.

In such cases, you need to make sure that the consumer stream's kernels (in this case, the D2H copy) before any kernels in the producer stream reuse that memory.

The idea then is to hold a reference to `p_cuda.grad` until the CPU has issued the ops with which you want the D2H copy to overlap with, and then you do `torch.cuda.current_stream().wait_event(event)` where `event` was recorded in `self.stream` right after the D2H copy and `current_stream()` is the default/producer stream. That way, any subsequent ops in the producer stream will run after the D2H copy has finished and can safely reuse the `p_cuda.grad` address.

Reply:
The challenge can be that you do not know how many / which ops to overlap with, so it is not convenient to sync back (`torch.cuda.current_stream().wait_event(d2h_event)`).

However, for cases like FSDP, we do have a good time: e.g. the previous reduce-scatter must finish before the next reduce-scatter, so let us wait for the previous reduce-scatter (doing this "sync back") right before the next reduce-scatter.

Reply:
`torch.cuda.current_stream().wait_event(event)` means that the next backward op cannot overlap with D2H?

Reply:
I think another option is to delete `p_cuda.grad` reference inside `optim.step()`, but it means we only start deallocating CUDA grad when we iterate over `self.queue` -> might not reduce much peak memory.

Reply:
> `torch.cuda.current_stream().wait_event(event)` means that the next backward op cannot overlap with D2H?

Yes, the next backward op after the `wait_event` call cannot overlap with the D2H right before the recorded event.

One way to think about it is to think about the actual CUDA address. The GPU gradient must have some address `A`. We need to make sure that no other op uses `A` until the D2H finishes. We can reserve `A` and make sure no other backward ops use it as long as we keep a reference to `A` (as a PyTorch implementation fact). At some point, we have overlapped enough ops, and we can then free `A`, requiring the aforementioned sync back.

Note that with `record_stream`, the address `A` will be reserved until the D2H copy finishes on GPU, at which point maybe many or even all backward ops were issued (in the most extreme case). In that case, none of the backward ops can actually reuse `A`. This memory reuse depends on the relative timing of CPU and GPU, which makes it difficult to reason about precisely.

Reply:
> I think another option is to delete `p_cuda.grad` reference inside `optim.step()`, but it means we only start deallocating CUDA grad when we iterate over `self.queue` -> might not reduce much peak memory.

I think calling `record_stream` probably dominates that approach because you will have to block CPU until the D2H finishes anyway.

Reply:
I see. Thank you for your detailed explanation!